hal/src/rtl872x/exflash_hal.cpp

```diff
@@ -400,32 +665,62 @@ int hal_exflash_write_special(hal_exflash_special_sector_t sp, uintptr_t addr, c
     CHECK_TRUE(sp == HAL_EXFLASH_SPECIAL_SECTOR_OTP, SYSTEM_ERROR_INVALID_ARGUMENT);
     CHECK_TRUE(data_buf && data_size > 0, SYSTEM_ERROR_INVALID_ARGUMENT);
 
-    // Read the first word at 0x00000000
-    uint32_t normalContent = 0;
-    FLASH_RxData(0, (uint32_t)0, sizeof(normalContent), (uint8_t*)&normalContent);
-    {
-        SCOPE_GUARD({
-            // Just in case, even if it might have failed to enter the secure OTP mode.
-            FLASH_TxCmd(MXIC_FLASH_CMD_EXSO, 0, nullptr);
+    if (flash_init_para.FLASH_Id == FLASH_ID_GD) {
+        return gd25PerformSecurityRegistersOperation(gd25WriteSecurityRegisters, addr, (uint8_t*)data_buf, data_size);
+    } else if (flash_init_para.FLASH_Id == FLASH_ID_MXIC) {
+        // Read the first word at 0x00000000
+        uint32_t normalContent = 0;
+        FLASH_RxData(0, (uint32_t)0, sizeof(normalContent), (uint8_t*)&normalContent);
+        {
+            SCOPE_GUARD({
+                // Just in case, even if it might have failed to enter the secure OTP mode.
+                FLASH_TxCmd(MXIC_FLASH_CMD_EXSO, 0, nullptr);
+                if (isSecureOtpMode(normalContent)) {
+                    SPARK_ASSERT(false);
+                }
+            });
+
+            FLASH_TxCmd(MXIC_FLASH_CMD_ENSO, 0, nullptr);
             if (isSecureOtpMode(normalContent)) {
-                SPARK_ASSERT(false);
+                CHECK(hal_flash_common_write(addr, data_buf, data_size, &perform_write, &hal_flash_common_dummy_read));
+            } else {
+                return SYSTEM_ERROR_INVALID_STATE;
             }
-        });
-
-        FLASH_TxCmd(MXIC_FLASH_CMD_ENSO, 0, nullptr);
-        if (isSecureOtpMode(normalContent)) {
-            CHECK(hal_flash_common_write(addr, data_buf, data_size, &perform_write, &hal_flash_common_dummy_read));
-        } else {
-            return SYSTEM_ERROR_INVALID_STATE;
         }
+    } else {
+        return SYSTEM_ERROR_NOT_SUPPORTED;
     }
     return SYSTEM_ERROR_NONE;
 }
 
 #endif // MODULE_FUNCTION != MOD_FUNC_BOOTLOADER
 
 int hal_exflash_erase_special(hal_exflash_special_sector_t sp, uintptr_t addr, size_t size) {
-    /* We only support OTP sector and since it's One Time Programmable, we can't erase it */
+    if (flash_init_para.FLASH_Id == FLASH_ID_GD) {
+        ExFlashLock lk(false); // Stop thread scheduler
+        // Write/Erase enable
+        FLASH_TxCmd(GENERIC_FLASH_CMD_WREN, 0, nullptr);
+        FLASH_WaitBusy(WAIT_WRITE_EN);
+
+        size_t otpPage = addr / HAL_QSPI_FLASH_OTP_SECTOR_SIZE_GD25;
+        if (otpPage > 2) {
+            return SYSTEM_ERROR_INVALID_ARGUMENT;
+        }
+        uint32_t address = ((otpPage + 1) << 12);
+
+        // Erase security registers
+        uint8_t buf[3] = {};
+        buf[0] = (uint8_t)((address >> 16) & 0xFF);
+        buf[1] = (uint8_t)((address >> 8) & 0xFF);
+        buf[2] = (uint8_t)((address >> 0) & 0xFF);
+        FLASH_TxCmd(GD_FLASH_CMD_ERS_SEC, sizeof(buf), buf);
```

Comment:
Do we even want to implement the ability to erase GD OTP data?

Reply:
I would leave it there. Because it's not our first time that the OTP info is programmed with wrong data in factory. With MXIC flash we have to replace the flash with a brand new one. But with GD flash it gives us a chance to re-program the OTP info with reworking the hardware.